### PR TITLE
feat: flame graph with WebGL

### PR DESCRIPTION
**Pull request by monfera**
## Summary

Status: 

- [x] API (column vectors)
- [x] scaling to OOM 1m nodes
- [x] click to focus
- [x] reset (click to empty area)
- [x] show adjacent nodes on the left and right side for context
- [x] text placement into visible part even if box is partly out of view
- [x] text truncation with no ellipsis to preserve info
- [x] proportional fonts to maximize visible text length
- [x] styling (color, gap, rounding)
- [x] API for passing two different datasets for tweening between them (for baseline use, `position0`/`position1` etc. should get the reference to the same typed array)
- [x] horizontal zoom (other than click to focus) with drag and cmd-wheel
- [x] vertical zoom (other than click to focus)
- [x] xy zoom/pan
- [x] specify row height, instead of fitting all rows into the available vertical space
- [x] wheel zooming with the mouse rather than touchpad interacts with the tooltip and highlight: eliminate the flashing 
- [x] fix for the mid-tween chart exit animation glitch
- [x] avoid text flashing on resize
- [x] rerender (text and box sharpening) upon changing display resolution or dragging from one display to another (changing devicePixelRatio)
- [x] rerender (text and box edge sharpening) upon exercising a browser pinch zoom
- [x] variable font size (supralinear with box height)
- [x] provide a margin around the chart, or some other means for the ability to zoom from the side exactly (@markov00's  suggestion)—achieved by zooming from edge if cursor is within 16 CSS pixels of the edge
- [x] to avoid a jump glitch, reset drag start when mouse exits the chart mid drag (short term); allow drag to continue to work outside the chart (medium term)
- [x] deactivate the hover tooltip over the minimap area
- [x] ~~open new issue for a future PR: add render counter and debug info onto DOM element attributes; accessibilitiy with keyboard navigation for all interactions~~ added https://github.com/elastic/elastic-charts/issues/1671

### Checklist

- [x] The proper **chart type** label has been added (e.g. `:xy`, `:partition`)
- [x] The proper **feature** labels have been added (e.g. `:interactions`, `:axis`)
- [ ] All related issues have been linked (i.e. `closes #123`, `fixes #123`)
- [x] New public API exports have been added to `packages/charts/src/index.ts`
- [x] The proper documentation and/or storybook story has been added or updated
- [x] The code has been checked for cross-browser compatibility (Chrome, Firefox, Safari, Edge)

### CODE REVIEW

File: packages/charts/api/charts.api.md [hunk 1]
```diff
@@ -555,6 +564,9 @@ export function computeRatioByGroups<T extends Record<string, unknown>>(data: T[
 // @public (undocumented)
 export type ContinuousDomain = [min: number, max: number];
 
+// @public
+export type ControlProviderCallback = (controlName: string, controlFunction: (...args: unknown[]) => void) => void;
```
Comment (monfera) [inline packages/charts/api/charts.api.md:568]: Not yet used in the code

File: packages/charts/api/charts.api.md [hunk 2]
```diff
@@ -57,6 +57,14 @@ export interface AngleFromTo {
 // @public
 export type AnimatedValue = number;
 
+// @public (undocumented)
+export interface AnimationConfig {
```
Comment (monfera) [inline packages/charts/api/charts.api.md:61]: Just an extraction

File: packages/charts/api/charts.api.md [hunk 3]
```diff
@@ -952,6 +964,39 @@ export type FitConfig = {
...
+    columnarData: ColumnarViewModel;
+    // (undocumented)
+    controlProviderCallback: ControlProviderCallback;
+    // (undocumented)
+    specType: typeof SpecType.Series;
+    // (undocumented)
+    valueAccessor: ValueAccessor<D>;
+    // (undocumented)
+    valueFormatter: ValueFormatter;
+    // Warning: (ae-forgotten-export) The symbol "ValueGetterFunction" needs to be exported by the entry point index.d.ts
```
Comment (monfera) [inline packages/charts/api/charts.api.md:994]: More work is needed for actual formatting, see todo

File: packages/charts/api/charts.api.md [hunk 4]
```diff
@@ -1662,7 +1708,7 @@ export type PartialTheme = RecursivePartial<Theme>;
 // Warning: (ae-forgotten-export) The symbol "PartitionSpec" needs to be exported by the entry point index.d.ts
 //
 // @public
-export const Partition: <D extends BaseDatum = any>(props: SFProps<PartitionSpec<D>, "chartType" | "specType", "animation" | "layout" | "valueAccessor" | "clockwiseSectors" | "specialFirstInnermostSector" | "drilldown" | "maxRowCount" | "fillOutside" | "radiusOutside" | "fillRectangleWidth" | "fillRectangleHeight" | "valueFormatter" | "valueGetter" | "percentFormatter" | "topGroove" | "smallMultiples" | "layers", never, "data" | "id">) => null;
+export const Partition: <D extends BaseDatum = any>(props: SFProps<PartitionSpec<D>, "chartType" | "specType", "animation" | "layout" | "valueAccessor" | "valueFormatter" | "valueGetter" | "clockwiseSectors" | "specialFirstInnermostSector" | "drilldown" | "maxRowCount" | "fillOutside" | "radiusOutside" | "fillRectangleWidth" | "fillRectangleHeight" | "percentFormatter" | "topGroove" | "smallMultiples" | "layers", never, "data" | "id">) => null;
```
Comment (monfera) [inline packages/charts/api/charts.api.md:1711]: Looks like a reorder

File: packages/charts/api/charts.api.md [hunk 5]
```diff
@@ -2249,6 +2295,8 @@ export interface ShapeTreeNode extends TreeNode, SectorGeomSpecY {
     // (undocumented)
     value: number;
     // (undocumented)
+    vmIndex: number;
```
Comment (monfera) [inline packages/charts/api/charts.api.md:2298]: I'll remove the added `vmIndex` props as vestigial

File: packages/charts/src/common/event_handler_selectors.ts [hunk 6]
```diff
@@ -50,15 +52,16 @@ export const getOnElementOutSelector = (prev: { pickedShapes: number | null }) =
   if (!settings.onElementOut) {
     return;
   }
-  const nextPickedShapes = pickedShapes.length;
+  const nextPickedShapes =
+    pickedShapes.length > 0 && pickedShapes[0].length > 0 && pickedShapes[0][0].vmIndex !== 0 ? 1 : 0;
```
Comment (markov00) [inline packages/charts/src/common/event_handler_selectors.ts:56]: can you add a small comment on that one? since vmIndex is only used for flamegraph is difficult to understand what this is doing here in a `common` code

File: packages/charts/src/chart_types/flame_chart/flame_chart.tsx [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+  }
+
+  private inTween = (t: DOMHighResTimeStamp) =>
+    this.drilldown.timestamp + this.props.animationDuration + TWEEN_EPSILON_MS >= t;
+
+  componentDidMount = () => {
+    /*
+     * the DOM element has just been appended, and getContext('2d') is always non-null,
+     * so we could use a couple of ! non-null assertions but no big plus
+     */
```
[the remark this answers is not in the document]
Comment (monfera) [reply packages/charts/src/chart_types/flame_chart/flame_chart.tsx:141]: It's one of the millions of TypeScript blind spots, which is why I added the comment. In a future TS version, maybe it'll be smart enough. The web standards guarantee that it be a non-null

File: packages/charts/src/chart_types/flame_chart/flame_chart.tsx [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+    e.stopPropagation();
+    const hovered = this.getHoveredDatumIndex(e);
+    const prevHoverIndex = this.hoverIndex >= 0 ? this.hoverIndex : NaN; // todo instead of translating NaN/-1 back and forth, just convert to -1 for shader rendering
+    if (hovered) {
+      this.hoverIndex = Number.isNaN(hovered.datumIndex) ? DUMMY_INDEX : hovered.datumIndex;
+      if (Number.isFinite(hovered.datumIndex) && !Object.is(this.hoverIndex, prevHoverIndex)) {
+        this.props.onElementOver([{ vmIndex: hovered.datumIndex }]); // userland callback
+      }
+      this.setState({});
+      this.drawCanvas(); // todo use setState properly which would also trigger drawCanvas
```
[the remark this answers is not in the document]
Comment (monfera) [reply packages/charts/src/chart_types/flame_chart/flame_chart.tsx]: Thanks Marco, good call! It used to be separate (to allow component change like tooltip removal) but not anymore. Will avoid the duplication

File: packages/charts/src/chart_types/flame_chart/flame_chart.tsx [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+    animationDuration: flameSpec?.animation.duration ?? 0,
+    chartDimensions: state.parentDimensions,
+    a11ySettings: getA11ySettingsSelector(state),
+    tooltipRequired: getTooltipType(settingsSpec) !== TooltipType.None,
+
+    // mandatory charts API protocol; todo extract these mappings once there are other charts like Flame
+    onElementOver: settingsSpec.onElementOver ?? (() => {}),
+    onElementClick: settingsSpec.onElementClick ?? (() => {}),
+    onElementOut: settingsSpec.onElementOut ?? (() => {}),
+    onRenderChange: settingsSpec.onRenderChange ?? (() => {}), // todo eventually also update data props on a local .echChartStatus element: data-ech-render-complete={rendered} data-ech-render-count={renderedCount} data-ech-debug-state={debugStateString}
```
[the remark this answers is not in the document]
Comment (monfera) [reply packages/charts/src/chart_types/flame_chart/flame_chart.tsx:466]: Indeed! I meant, it'll be needed for Lens and similar apps but afaik isn't blocking the initial Observability use case

File: packages/charts/src/chart_types/flame_chart/render/animated_draw.ts [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { LabelAccessor } from '../../../utils/common';
+import { AnimationState, ContinuousDomainFocus, GLResources } from '../types';
+import { drawFrame } from './draw_a_frame';
+
+const linear = (x: number) => x;
+const easeInOut = (alpha: number) => (x: number) => x ** alpha / (x ** alpha + (1 - x) ** alpha);
```
[the remark this answers is not in the document]
Comment (monfera) [reply packages/charts/src/chart_types/flame_chart/render/animated_draw.ts]: History: it comes from the standalone prototype, and I'm planning to replace tweening with a recurrence relation, so it'll be gone

File: packages/charts/src/chart_types/flame_chart/render/draw_canvas.ts [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+      ctx.beginPath();
+      ctx.rect(x, y - zoomedRowHeight * cssHeight, width, zoomedRowHeight * cssHeight);
+      if (textColor !== lastTextColor) {
+        // as we're sorting the iteration, the number of color changes (API calls) is minimized
+        ctx.fillStyle = textColor;
+        lastTextColor = textColor;
+      }
+      ctx.save();
+      ctx.clip();
+      ctx.fillText(label, x + TEXT_PAD_LEFT, y - ROW_OFFSET_Y * rowHeight * cssHeight);
```
[the remark this answers is not in the document]
Comment (monfera) [reply packages/charts/src/chart_types/flame_chart/render/draw_canvas.ts:73]: The `maxWidth` argument ensures that the text isn't spilling out of the specified width, but it does so by shrinking the text 🤣 Which makes it a strong contender for the "least uselful element of the Web API" 😄

[the remark this answers is not in the document]
Comment (markov00) [reply packages/charts/src/chart_types/flame_chart/render/draw_canvas.ts:73]: oh really? ...that so bad!

File: packages/charts/src/chart_types/flame_chart/render/ensure_webgl.ts [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+  const readPixelXY: PickFunction = (x, y) => {
+    if (gl) {
+      bindFramebuffer(gl, GL_READ_FRAMEBUFFER, pickTexture.target());
+      const pixel = readPixel(gl, dpr * x, textureHeight - dpr * y);
+      const found = pixel[0] + pixel[1] + pixel[2] + pixel[3] > 0;
+      const datumIndex = found
+        ? pixel[3] + 256 * (pixel[2] + 256 * (pixel[1] + 256 * pixel[0])) - GEOM_INDEX_OFFSET
+        : NaN;
+      return Number.isNaN(datumIndex) ? NaN : datumIndex;
```
[the remark this answers is not in the document]
Comment (monfera) [reply packages/charts/src/chart_types/flame_chart/render/ensure_webgl.ts]: 4559c0a1d210a559df1703fc7e417c0dde3d9ee3

File: packages/charts/src/chart_types/flame_chart/flame_chart.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (monfera) [reply packages/charts/src/chart_types/flame_chart/flame_chart.tsx]: 4af121e01c6ff1f496fbae241d96e1a525c3b731

File: packages/charts/src/chart_types/flame_chart/flame_api.ts [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    keyof typeof buildProps['overrides'],
+    keyof typeof buildProps['defaults'],
+    keyof typeof buildProps['optionals'],
+    keyof typeof buildProps['requires']
+  >,
+) {
+  const { defaults, overrides } = buildProps;
+  useSpecFactory<FlameSpec<D>>({ ...defaults, ...stripUndefined(props), ...overrides });
+  return null;
+};
```
Comment (nickofthyme) [inline packages/charts/src/chart_types/flame_chart/flame_api.ts:66]: You mentioned this being a little complex and you are correct 😅. The simpler approach is to use `specComponentFactory` directly which simply takes the overrides and defaults and handles the rest for you.

However, using `specComponentFactory` does not allow you to use a top-level generic type, hence the crazy type magic. In this case you are using `D` to pass as a generic to `FlameSpec`, if that is not needed you can you the simpler approach.

I am also now noticing that the inside call the `useSpecFactory` could be simplified a little.

File: packages/charts/src/chart_types/flame_chart/flame_chart.tsx [hunk 14]
```diff
@@ -0,0 +1,598 @@
...
+    this.pickTexture = NullTexture;
+    this.glResources = NULL_GL_RESOURCES;
+    this.glCanvasRef = createRef();
+    this.animationRafId = NaN;
+    this.prevT = NaN;
+    this.currentFocus = focusRect(this.props.columnarViewModel, props.chartDimensions.height, 0, -Infinity);
+    this.targetFocus = { ...this.currentFocus };
+    this.hoverIndex = NaN;
+    this.pointerX = -10000;
+    this.pointerY = -10000;
```
Comment (nickofthyme) [inline packages/charts/src/chart_types/flame_chart/flame_chart.tsx:176]: What is the significance of this value? And not being `Infinity`?

File: packages/charts/src/chart_types/flame_chart/flame_chart.tsx [hunk 15]
```diff
@@ -0,0 +1,598 @@
...
+
+  componentWillUnmount() {
+    this.props.containerRef().current?.removeEventListener('wheel', this.preventScroll);
+  }
+
+  componentDidUpdate = () => {
+    if (!this.ctx) this.tryCanvasContext();
+    this.ensurePickTexture();
+    this.drawCanvas(); // eg. due to chartDimensions (parentDimensions) change
+    // this.props.onChartRendered() // creates and infinite update loop
```
Comment (nickofthyme) [inline packages/charts/src/chart_types/flame_chart/flame_chart.tsx:232]: @monfera could you take a quick look at this to see what could fix this? Not a blocker but just wanted to make you aware of it.

File: packages/charts/src/chart_types/flame_chart/render/draw_canvas.ts [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+  const minRectWidthForTextInCssPix = minTextLengthCssPix + TEXT_PAD_LEFT + TEXT_PAD_RIGHT;
+  const minRectWidth = minRectWidthForTextInCssPix / cssWidth;
+  const textColor = 'black'; // todo it could come from config / theme or automatic decision like in other charts
+
+  // text rendering
+  ctx.save();
+  ctx.textAlign = 'left';
+  ctx.textBaseline = 'middle';
+  ctx.scale(dpr, dpr);
+  ctx.font = `${fontSize}px "Atkinson Hyperlegible"`;
```
Comment (nickofthyme) [inline packages/charts/src/chart_types/flame_chart/render/draw_canvas.ts:52]: Do we want to hardcode this?

File: packages/charts/src/chart_types/partition_chart/renderer/dom/highlighter.tsx [hunk 17]
```diff
@@ -282,8 +282,7 @@ export const DEFAULT_PROPS: HighlighterProps = {
 /** @internal */
 export function highlightSetMapper(geometries: QuadViewModel[], foci: IndexedContinuousDomainFocus[]) {
   return (vm: ShapeViewModel): HighlightSet => {
-    const { index } = vm;
-    const { innerIndex } = vm;
+    const { index, innerIndex } = vm;
```
Comment (nickofthyme) [inline packages/charts/src/chart_types/partition_chart/renderer/dom/highlighter.tsx:285]: 😌

File: packages/charts/src/common/kingly.ts [hunk 18]
```diff
@@ -0,0 +1,626 @@
+/*
```
Comment (nickofthyme) [inline packages/charts/src/common/kingly.ts:1]: Once we get the package logic sorted to where we can have multiple packages, this seems like a good candidate to be its own package.

File: storybook/main.js [hunk 19]
```diff
@@ -10,7 +10,7 @@ module.exports = {
   stories: ['./stories/**/*.(stories|story).ts?(x)'],
   addons: [
     '@storybook/addon-actions',
-    '@storybook/addon-docs',
+    // '@storybook/addon-docs',
```
Comment (nickofthyme) [inline storybook/main.js:13]: What was this issue this was posing?

File: storybook/style.scss [hunk 20]
```diff
@@ -127,7 +127,7 @@ body {
   position: relative;
   z-index: 500;
   box-sizing: border-box;
-  background-color: blanchedalmond;
+  /* background-color: blanchedalmond; */
```
Comment (nickofthyme) [inline storybook/style.scss:130]: 😱 So is this official @markov00?

**Review by nickofthyme — APPROVED** [review]
Looks great Robert!

Thanks for walking me through the code at a top-level that helped me wrap my head around this code much easier.

I tested locally and had the chart disappear on me once when clicking around but haven't been able to reproduce it. Also turned off webGL and the difference is night 🌃  and day 🏙️!

File: storybook/stories/bar/54_functional_accessors.story.tsx [hunk 21]
```diff
@@ -31,11 +32,16 @@ export const Example = () => {
   const splitAccessorFn: AccessorFn = (d) => d.g2;
   splitAccessorFn.fieldName = text('split fn name', '') || undefined;
 
-  const onElementClick: ElementClickListener = ([[, { key }]]) => action('clicked series key')(key);
+  const onElementClick = ([[, { key }]]: XYChartElementEvent[]) => action('clicked series key')(key);
```
Comment (nickofthyme) [inline storybook/stories/bar/54_functional_accessors.story.tsx:35]: @markov00 We need to come up with a better API for this, the union types are starting to diverge too much. I think a simple fix would just be to change the events array to a record per type, so like...

```
interface ClickEventElementMap {
  xy: XYChartElementEvent[];
  partition: PartitionElementEvent[];
  flame: FlameElementEvent[];
  heatmap: HeatmapElementEvent[];
  wordCloud: WordCloudElementEvent[];
}
export type ElementClickListener = (elements: ClickEventElementMap) => void;
```

Not saying in this PR!

File: storybook/style.scss (hunk 20 above)
Comment (monfera) [reply storybook/style.scss:130]: I pitched this in our channel a couple of days ago, not sure yet where Marco stands on this issue. The upshot is that the blanchedalmond seems to have served its purpose, and feels like it's no longer worth having it, due to the visual interference it causes. It snuck into the PR by accident. Happy to either remove the line or remove the asterisks. Alternatively we could put a 1px grey border around all the charts (like the WebGL flame, or fainter) which is unlikely to be mistaken with chart rendering

File: packages/charts/src/chart_types/flame_chart/flame_chart.tsx (hunk 14 above)
Comment (monfera) [reply packages/charts/src/chart_types/flame_chart/flame_chart.tsx:176]: It was just a copy-paste offscreen value from someplace else at a time when switching it would've been tangential. Will revise

File: packages/charts/src/chart_types/flame_chart/flame_chart.tsx (hunk 15 above)
Comment (monfera) [reply packages/charts/src/chart_types/flame_chart/flame_chart.tsx:232]: The component update triggers a redux action(!) and that eventually cascades into another component update. See our channel convo with Marco on May 2 to my question earlier (search for `CHART_RENDERED`)`. This circularity was the motivation. 

Upshot: I feel like the proper fix would be the demotion of the `CHART_RENDERED` action, even if the current circularity could be technically fixed. Because it is not an action (open to prepared arguments). We seemed to arrive at a consensus that over time its action-ness could be phased out. As there's no current need for this callback for this chart, I didn't want to spend time fixing it for the old way, and doing the new way, across all chart types, is not for this PR

File: packages/charts/src/chart_types/flame_chart/render/draw_canvas.ts (hunk 16 above)
Comment (monfera) [reply packages/charts/src/chart_types/flame_chart/render/draw_canvas.ts:52]: Oh totally. Not. Even if I love it on accessibility grounds. I put this on the todo list in the Description above, probably for a follow-up PR.   

Even when it happens, I'd like to keep using this font for this mock because it's a disambiguated font, here are my [couple of lines](https://github.com/elastic/elastic-charts/issues/1365) on this topic (maybe the Canvas2d handling is caniuse-able by now?)

**Comment by nickofthyme** [comment]
I updated the puppeteer screenshots from my intel mac (https://github.com/elastic/elastic-charts/pull/1664/commits/77e78f04b6bf606311625b79ac4041bbd865a0c4). There are 2 stories with errors in code preventing them from rendering, see:

1. https://ech-e2e-ci--pr-1664-8xu76vsj.web.app/?path=/story/sunburst--no-slice
2. https://ech-e2e-ci--pr-1664-8xu76vsj.web.app/?path=/story/sunburst--total-zero

File: storybook/main.js (hunk 19 above)
Comment (monfera) [reply storybook/main.js:13]: Marco made a mock with 110k data points, which we use for testing, but it's not in the repo due to its size. I can send it to you if you want to try it. There's load time slowdown if the addon-docs remains active. Though the `//` should probably be deleted before merge, lmk

File: storybook/stories/bar/54_functional_accessors.story.tsx (hunk 21 above)
Comment (monfera) [reply storybook/stories/bar/54_functional_accessors.story.tsx:35]: We could also have a broader chat on what my mind calls "elastic-charts protocol". The intent is a level of uniformity, which is sensible. Yet it's also a set of somewhat vaguely delineated internal conventions that don't have documented reasons afaik. For some of these things, [composition > convention](https://clojurefun.wordpress.com/2012/08/17/composition-over-convention/).

File: packages/charts/src/chart_types/flame_chart/flame_api.ts (hunk 13 above)
Comment (monfera) [reply packages/charts/src/chart_types/flame_chart/flame_api.ts:66]: Most of the broilerplate and maybe some other few hundred lines wouldn't be needed if our entire spec lived in JSON, right? Complexity whose main purpose is to make Charts appear as a React component with other Charts React components as children. If I understand, it has no technical or functional purpose, it's cosmetics, which was useful at the beginning but might no longer be a net positive. IIRC it's not just delivered and executed code, type complications etc. but there are some consequences like only probabilistically knowing we got to the end of spec processing

tl;dr when seeing `...Factory`, I start to wonder if the root cause could be addressed. I'd like to learn about current or future Charts use cases of anything with `Factory` in its name where it is fundamental or beneficial for the end user, rather than as a way of addressing implementation choices

File: packages/charts/src/common/event_handler_selectors.ts (hunk 6 above)
Comment (monfera) [reply packages/charts/src/common/event_handler_selectors.ts]: I reverted the `vmIndex` for now in f3b61427

File: packages/charts/src/chart_types/flame_chart/flame_api.ts (hunk 13 above)
Comment (monfera) [reply packages/charts/src/chart_types/flame_chart/flame_api.ts:66]: Deep linked this convo in https://github.com/elastic/elastic-charts/issues/1671 so this can be marked as resolved

File: packages/charts/src/chart_types/flame_chart/flame_chart.tsx (hunk 15 above)
Comment (monfera) [reply packages/charts/src/chart_types/flame_chart/flame_chart.tsx:235]: Deep linked it in https://github.com/elastic/elastic-charts/issues/1671 so this can be resolved

File: packages/charts/src/chart_types/flame_chart/render/draw_canvas.ts (hunk 16 above)
Comment (monfera) [reply packages/charts/src/chart_types/flame_chart/render/draw_canvas.ts:52]: Deep linked it in https://github.com/elastic/elastic-charts/issues/1671 so it can be resolved

File: storybook/stories/bar/54_functional_accessors.story.tsx (hunk 21 above)
Comment (monfera) [reply storybook/stories/bar/54_functional_accessors.story.tsx:35]: I deep linked it from https://github.com/elastic/elastic-charts/issues/1671 so it can be marked as resolved here

File: .prettierignore [hunk 22]
```diff
@@ -10,3 +10,5 @@ coverage/
 **/tmp/
```
Comment (markov00) [inline .prettierignore:10]: Two VRTs comments:
- there are two snapshots that should be removed: the pie charts with the `2-snap` suffix
- any ideas about this 0 bytes change for the annotation chart?

**Review by markov00 — APPROVED** [review]
It looks awesome! Great job Robert!
I've tested it locally on mac on Safari, Chrome and Firefox. Everything works flawlessly and smoothly.
I've added a small possible feature to #1671 an alternative to the minimap with scrollbars, just and idea, and we can discuss in a later stage

File: packages/charts/src/chart_types/flame_chart/flame_chart.tsx [hunk 23]
```diff
@@ -0,0 +1,638 @@
...
+
+  private datumAtXY: PickFunction = (x, y) =>
+    this.glContext ? colorToDatumIndex(readPixel(this.glContext, x, y)) : NaN;
+
+  private updatePointerLocation(e: { clientX: number; clientY: number }) {
+    if (!this.props.forwardStageRef.current || !this.ctx) return;
+    const box = this.props.forwardStageRef.current.getBoundingClientRect();
+    this.pointerX = e.clientX - box.left;
+    this.pointerY = e.clientY - box.top;
+  }
```
[the remark this answers is not in the document]
Comment (monfera) [reply packages/charts/src/chart_types/flame_chart/flame_chart.tsx:246]: Thanks, I added it to the new issue

File: packages/charts/src/chart_types/flame_chart/flame_chart.tsx [hunk 24]
```diff
@@ -0,0 +1,638 @@
...
+    const style: CSSProperties = {
+      width,
+      height,
+      top: 0,
+      left: 0,
+      padding: 0,
+      margin: 0,
+      border: 0,
+      position: 'absolute',
+      cursor: this.hoverIndex >= 0 ? 'pointer' : DEFAULT_CSS_CURSOR,
```
[the remark this answers is not in the document]
Comment (monfera) [reply packages/charts/src/chart_types/flame_chart/flame_chart.tsx:455]: Great idea, added it to the issue, though maybe it can catch this PR too

**Comment by monfera** [comment]
@markov00 addon-docs reactivated in b7fe94bf

File: packages/charts/src/chart_types/flame_chart/flame_chart.tsx (hunk 24 above)
[the remark this answers is not in the document]
Comment (monfera) [reply packages/charts/src/chart_types/flame_chart/flame_chart.tsx:455]: Added it 50940dc5

File: .prettierignore (hunk 22 above)
Comment (markov00) [reply .prettierignore:10]: solved in https://github.com/elastic/elastic-charts/pull/1664/commits/568f7434540bda44e018dafa8cfcbacc7da2fbc1
